File: linked list problems/Java/DLL.java
import java.util.*;  


class Node {

	int data;
	Node next;
	Node prev;

	Node(int data)
	{
		this.data = data;
		this.next = null;
		this.prev=null;
	}

}

class DLL
{
	public static void main(String args[])
	{
		Node head = create_DLL();
		printLinkedlist(head);

		Node test = insert_pos(head);
		printLinkedlist(test);
		Node test1 = delete_pos(head);
		printLinkedlist(test1);
		Node test2 = reverse(head);
		printLinkedlist(test2);
	}

	public static Node create_DLL()
	{
		Scanner sc = new Scanner(System.in);
		Node head = null;
		boolean done = false;
		System.out.println("please enter the data to create a DLL ");
		while(done==false)
		{

			String input = sc.next();
			if(input.equals("done"))
				break;
			int data = Integer.parseInt(input);
			Node new_node = new Node(data);

			if (head == null)
			{
				head = new_node;
			}
			else
			{
				Node temp = head;
				while(temp.next!=null)
					temp = temp.next;

				new_node.prev = temp;
				temp.next = new_node;
			}

			System.out.println("New node "+ data +" is added to the linked list and type done to stop it");
		}

		return head;

	}

	public static void printLinkedlist(Node head)
	{
		Node temp = head;
		Node last = null;
		System.out.println("---------------");
		while(temp!=null)
		{
			last = temp;
			System.out.print(temp.data+ " ");
			temp = temp.next;
		}
		System.out.println();

		while(last!=null)
		{
			System.out.print(last.data+ " ");
			last = last.prev;
		}
		System.out.println();
		System.out.println("---------------");

	}

	public static Node insert_pos(Node head)
	{
		Scanner sc = new Scanner(System.in);

		System.out.println("Please enter the data to insert ");
		int data = sc.nextInt(); 
		System.out.println("Plese enter the index where to be inserted");
		int pos = sc.nextInt();
		
		Node new_node = new Node(data);
		Node temp = head;

		int curr_pos = 1;

		if (pos == 0)
		{
			new_node.next = head;
			head.prev = new_node;
			head = new_node;
			return head;
		}

		while(temp.next!=null)
		{
			if (curr_pos== pos)
			{
				new_node.next = temp.next;
				temp.next.prev = new_node;
				new_node.prev = temp;
				temp.next = new_node;
				System.out.println("The data has been inserted successfully");
				return head;
			}

			temp= temp.next;
			curr_pos+=1;

			if (temp.next==null && curr_pos==pos)
			{
				temp.next = new_node;
				new_node.prev = temp;
				System.out.println("The data has been inserted successfully");
				return head;
			}
		}

		System.out.println("The index is out of range or invalid");
		return head;
	}

	public static Node delete_pos(Node head)
	{
		System.out.println("Please enter the index where the node should be deleted");
		Scanner sc = new Scanner(System.in);
		int pos = sc.nextInt();

		int curr_pos = 1;

		if (pos == 0)
		{
			head = head.next;
			head.prev = null;
			System.out.println("The data has been deleted successfully");
			return head;
		}

		Node temp = head;

		while(temp.next!=null)
		{
			if (curr_pos==pos)
			{
				temp.next = temp.next.next;
				if (temp.next!=null)
					temp.next.prev = temp;
				System.out.println("The data has been deleted successfully");
				return head;
			}
			temp = temp.next;
			curr_pos +=1;

			if (temp.next==null && curr_pos==pos)
			{
				temp.prev.next = null;
				temp.prev = null;
				System.out.println("The data has been delteted successfully");
				return head;
			}
		}

		System.out.println("The index is out of range or invalid");
		return head;
	}

	public static Node reverse(Node head)
	{
		Node prev = null;
		Node next = null;
		Node curr = head;

		while(curr!=null)
		{
			next = curr.next;
			curr.next = curr.prev;
			curr.prev = next;
			prev = curr;
			curr = next;
		}

		head = prev;
		return head;
	}
}

File: linked list problems/Python/DLL.py

class Node ():

	def __init__(self,data):
		self.data = data
		self.next = None
		self.prev = None

def create_DLL():
	print("Please enter the data to create a DLL")
	done = False
	head = None

	while not done:
		data = input()

		if data =="done":
			break
		data = int(data)
		new_node = Node(data)

		if head == None:
			head = new_node
		else:
			temp = head

			while temp.next!=None:
				temp = temp.next
			
			temp.next = new_node
			new_node.prev = temp

		print("The data has been added type done to complete the list")

	return head


def printLinkedList(head):
	
	temp = head
	last = None
	print("-------------")
	while temp!=None:
		last = temp
		print(temp.data,"",end='')
		temp = temp.next
	print()

	while last!=None:
		print(last.data,"",end='')
		last = last.prev
	print()
	print("-------------")

def insert_pos(head):
	print("Please enter the data to be inserted")
	data = int(input())
	print("Please enter the index to be inserted")
	pos = int(input())

	curr_pos = 1

	new_node = Node(data)
	temp = head

	if pos==0:
		new_node.next = head
		head.prev = new_node
		head = new_node
		print("The node has been inserted at ",pos," index");
		return head

	while temp.next!=None:

		if curr_pos == pos:
			new_node.next = temp.next
			temp.next.prev = new_node
			temp.next = new_node
			new_node.prev = temp
			print("The node has been inserted at ",pos," index");
			return head
		temp = temp.next
		curr_pos +=1

		if temp.next==None and curr_pos==pos:
			temp.next = new_node
			new_node.prev = temp
			print("The node has been inserted at ",pos," index");
			return head

	print("index is out of range or invalid")
	return head

def delete_pos(head):
	print("please enter the index where the node to be deleted")
	pos = int(input())
	curr_pos =1

	if pos == 0:
		head.next.prev = None
		head = head.next
		print("The node has been deleted at ",pos," index");
		return head

	temp = head

	while temp.next!=None:

		if curr_pos ==pos:
			temp.next = temp.next.next
			if temp.next!=None:
				temp.next.prev = temp
			print("The node has been deleted at ",pos," index");
			return head
		
		temp = temp.next;
		curr_pos += 1

		if temp.next==None and curr_pos==pos:
			temp.prev.next = None
			temp.prev = None
			print("The node has been deleted at ",pos," index");
			return head

	print("index is out of range or invalid")
	return head

def reverse(head):

	prev, curr, next = None,head,None

	while curr!=None:
		next = curr.next
		curr.next = curr.prev
		curr.prev = next
		prev = curr
		curr = next

	head = prev

	return head


if __name__ == "__main__":
	head = create_DLL()
	printLinkedList(head)
	test = reverse(head)
	print("The reversed list is")
	printLinkedList(test)